p4include/v1model.p4

```diff
@@ -88,27 +88,123 @@ enum MeterType {
 
 extern counter {
     counter(bit<32> size, CounterType type);
+    /***
+     * count() causes the counter state with the specified index to be
+     * read, modified, and written back, atomically relative to the
+     * processing of other packets, updating the packet count, byte
+     * count, or both, depending upon the CounterType of the counter
+     * instance used when it was constructed.
+     *
+     * @param index The index of the counter state in the array to be
+     *              updated, normally a value in the range [0,
+     *              size-1].  If index >= size, no counter state will be
+     *              updated.
+     */
     void count(in bit<32> index);
 }
 
 extern direct_counter {
     direct_counter(CounterType type);
+    /***
+     * count() on a direct_counter object may only be called within an
+     * action executed as a result of matching a table entry, of a
+     * table that has a direct_counter associated with it.  Calling
+     * count() causes the counter state associated with the matched
+     * table entry to be read, modified, and written back, atomically
+     * relative to the processing of other packets.
+     *
+     * TBD: Is the direct_counter state updated even if an action is
```

Comment:
In bmv2 simple_switch, either a table has a direct counter or doesn't, and if it does the counter will be updated for every table lookup (that doesn't miss), irrespective of the action. I assume that the simple switch backend in p4c just looks at the presence / absence of the table property and doesn't verify that count is called in every action (although that would be a good check to have, and we could print a warning if count is not called in all actions).

Implementing PSA direct counters will require some work because the spec doesn't mandate that the counter state has to be updated for every action, but only for actions that call count. I'm hoping the Cornell team can get to it.

Reply:
Thanks.  I am adding another commit for the direct_counter constructor explaining how to associate a direct_counter instance to a table using the "counters = <object_name>;" table property, and changing the documentation to the read() method to indicate that it is really an unnecessary method in v1model, since the counter state for each matched table entry will be updated regardless of whether the count() method is called or not.